### PR TITLE
Don't fold `aten.detach` if result isn't same type as input.

### PR DESCRIPTION
**Pull request by AaronStGeorge**
We were seeing some assertion failures after some checks around folders were tightened up in LLVM: https://github.com/llvm/llvm-project/pull/75887 . This PR essentially moves the logic that used to be applied at the LLVM level into the folder, which seems to be the suggested fix.

I'm not sure if the IR that caused issues for us _should_ be valid?
```
%1 = torch.aten.detach %arg0 : !torch.tensor<[1],f32> -> !torch.tensor
```
A better fix might be to create a verifier ensuring the result of `aten.detach` has the same type as its operand.

### CODE REVIEW

File: test/Dialect/Torch/canonicalize.mlir [hunk 1]
```diff
@@ -2146,3 +2146,9 @@ func.func @torch.aten.masked_fill.Tensor$canonicalize(%arg0: !torch.vtensor<[?,?
   %1 = torch.aten.masked_fill.Tensor %arg0, %arg1, %0 : !torch.vtensor<[?,?],f32>, !torch.vtensor<[?,?],i1>, !torch.vtensor<[],f32> -> !torch.vtensor<[?,?],f32>
   return %1 : !torch.vtensor<[?,?],f32>
 }
+
+// CHECK-LABEL:   func.func @torch.aten.detach$canonicalize
```
Comment (sjain-stanford) [inline test/Dialect/Torch/canonicalize.mlir:2150]: Did you intend to add a `CHECK-NEXT` for `torch.aten.detach` (to confirm it didn't get folded by `-canonicalize`)?

**Review by sjain-stanford — APPROVED** [review]
LGTM
